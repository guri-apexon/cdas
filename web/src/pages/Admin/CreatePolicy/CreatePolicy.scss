@import '../../../styles/customscroll.scss';
.create-policy-wrapper{
    padding: 0 28px;
    .breadcrump{
        margin-left: -8px;
        margin-bottom: 0;
        padding-left: 0;
    }
    .top-actions{
        justify-content: end;
        label{
            margin-right: 40px;
        }
    }
    .create-sidebar{
        max-width: 75%;
        .title{
            font-size: 24px;
        }
    }
    .products-wrapper{
        margin-top: 10px;
        .product-content{
            margin-top: 20px;
        }
    }
    .permission-table-wrapper{
        position: relative;
        height: 80vh;
        @extend .customscroll;
        tfoot{
            display: none;
        }
        .search-header{
            position: absolute;
            right: 20px;
            top: 6px;
            z-index: 2;
        }
    }
}