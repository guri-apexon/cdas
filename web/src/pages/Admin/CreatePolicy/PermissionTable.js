/* eslint-disable no-prototype-builtins */
import React, { memo, useEffect, useState } from "react";
import Table, {
  createSelectFilterComponent,
  createStringSearchFilter,
  dateFilterV2,
  numberSearchFilter,
  compareDates,
  compareNumbers,
  compareStrings,
} from "apollo-react/components/Table";
import Search from "apollo-react/components/Search";

const CustomHeader = ({ setSearchText, searchTxt }) => {
  const [txt, setText] = useState(searchTxt);
  const setTextHandler = (e) => {
    setText(e.target.value);
  };
  const onKeyPress = (e) => {
    if (e.key === "Enter") {
      setSearchText(e.target.value);
    }
  };
  return (
    <>
      <Search
        placeholder="Search"
        size="small"
        value={txt}
        onKeyPress={onKeyPress}
        onChange={setTextHandler}
      />
    </>
  );
};

const PermissionTable = ({ title, data, updateData, messageContext }) => {
  const [tableRows, settableRows] = useState(data);
  const [filteredData, setFilteredData] = useState(data);
  const [searchTxt, setSearchText] = useState("");
  const FeatureCell = ({ row, column: { accessor } }) => {
    return <span className="b-font">{row[accessor]}</span>;
  };
  const handleChange = (e, row) => {
    const { checked, accessor } = e.target;
    const type = e.target.getAttribute("data-accessor");
    row.permsn_nm[type] = checked;
    switch (type) {
      case "Create":
        if (checked) {
          if (row.permsn_nm.hasOwnProperty("Read"))
            row.permsn_nm.Read = checked;
        }
        break;
      case "Update":
        if (row.permsn_nm.hasOwnProperty("Read")) row.permsn_nm.Read = checked;
        if (!checked) {
          if (row.permsn_nm.hasOwnProperty("Delete"))
            row.permsn_nm.Delete = checked;
        }
        break;
      case "Read":
        if (!checked) {
          messageContext.showErrorMessage(
            "Deselecting Read will automatically remove the options for all other options",
            null,
            "info"
          );
          if (row.permsn_nm.hasOwnProperty("Create"))
            row.permsn_nm.Create = false;
          if (row.permsn_nm.hasOwnProperty("Update"))
            row.permsn_nm.Update = false;
          if (row.permsn_nm.hasOwnProperty("Delete"))
            row.permsn_nm.Delete = false;
          if (row.permsn_nm.hasOwnProperty("Download"))
            row.permsn_nm.Download = false;
        }
        break;
      case "Delete":
        if (checked) {
          if (row.permsn_nm.hasOwnProperty("Read"))
            row.permsn_nm.Read = checked;
          if (row.permsn_nm.hasOwnProperty("Update"))
            row.permsn_nm.Update = checked;
        }
        break;
      case "Download":
        if (checked) {
          if (row.permsn_nm.hasOwnProperty("Read"))
            row.permsn_nm.Read = checked;
        }
        break;
      default:
        break;
    }
    const tableData = [
      ...new Set([...filteredData, ...tableRows].map((obj) => obj)),
    ];
    // settableRows(tableData);
    updateData({ product: title, data: tableData });
  };
  const checkboxCell = ({ row, column: { accessor } }) => {
    if (!row.permsn_nm.hasOwnProperty(accessor)) {
      return false;
    }
    return (
      <input
        type="checkbox"
        className="custom-checkbox"
        data-accessor={accessor}
        checked={row.permsn_nm[accessor]}
        onChange={(e) => handleChange(e, row)}
      />
    );
  };
  const columns = [
    {
      header: "Category",
      accessor: "ctgy_nm",
      sortFunction: compareStrings,
      width: 150,
    },
    {
      header: <span className="b-font">Features</span>,
      accessor: "feat_nm",
      sortFunction: compareStrings,
      customCell: FeatureCell,
    },
    {
      header: "Read",
      accessor: "Read",
      width: 100,
      customCell: checkboxCell,
    },
    {
      header: "Update",
      accessor: "Update",
      width: 100,
      customCell: checkboxCell,
    },
    {
      header: "Create",
      accessor: "Create",
      width: 100,
      customCell: checkboxCell,
    },
    {
      header: "Delete",
      accessor: "Delete",
      width: 100,
      customCell: checkboxCell,
    },
    {
      header: "Download",
      accessor: "Download",
      width: 100,
      customCell: checkboxCell,
    },
    {
      header: "Enable",
      accessor: "Enable",
      width: 100,
      customCell: checkboxCell,
    },
  ];
  const handleCheckox = (e, checked) => {
    console.log("columnName", checked);
  };
  useEffect(() => {
    const newRows = tableRows.filter((row) => {
      return row.feat_nm.toLowerCase().includes(searchTxt);
    });
    setFilteredData(newRows);
  }, [searchTxt]);
  useEffect(() => {
    console.log("table Updated");
    settableRows(data);
    if (filteredData.length === 0) setFilteredData(data);
  }, [data]);
  return (
    <div className="permission-table-wrapper">
      <Table
        title="Permissions"
        subtitle={title}
        columns={columns}
        rows={filteredData}
        rowsPerPage={tableRows.length}
        CustomHeader={() => (
          <CustomHeader searchTxt={searchTxt} setSearchText={setSearchText} />
        )}
        initialSortedColumn="feat_nm"
        initialSortOrder="asc"
      />
    </div>
  );
};
export default memo(PermissionTable);
