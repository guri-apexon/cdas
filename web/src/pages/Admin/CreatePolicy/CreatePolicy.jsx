/* eslint-disable react/button-has-type */
import React, { useContext, useEffect, useState } from "react";
import Box from "apollo-react/components/Box";
import { useHistory } from "react-router";
import BreadcrumbsUI from "apollo-react/components/Breadcrumbs";
import Switch from "apollo-react/components/Switch";
import ButtonGroup from "apollo-react/components/ButtonGroup";
import TextField from "apollo-react/components/TextField";
import "./CreatePolicy.scss";
import Typography from "apollo-react/components/Typography";
import Grid from "apollo-react/components/Grid";
import Tab from "apollo-react/components/Tab";
import Tabs from "apollo-react/components/Tabs";
import {
  addPolicyService,
  fetchProducts,
  getPolicyPermissions,
} from "../../../services/ApiServices";
import { MessageContext } from "../../../components/Providers/MessageProvider";
import PermissionTable from "./PermissionTable";
import { getUserInfo, inputAlphaNumeric } from "../../../utils";

const breadcrumpItems = [
  { href: "/" },
  {
    title: "Policy Management",
    href: "/policy-management",
  },
  {
    title: "Create New Policy",
  },
];

const CreatePolicy = () => {
  const [active, setActive] = useState(true);
  const [loading, setLoading] = useState(false);
  const [currentTab, setCurrentTab] = useState(0);
  const [policyName, setPolicyName] = useState("");
  const [policyDesc, setPolicyDesc] = useState("");
  const [permissions, setPermissions] = useState({});
  const [products, setProducts] = useState([]);
  const messageContext = useContext(MessageContext);
  const userInfo = getUserInfo();
  const history = useHistory();
  const handleActive = (e, checked) => {
    setActive(checked);
  };
  const handleChangeTab = (event, v) => {
    setCurrentTab(v);
  };
  // eslint-disable-next-line consistent-return
  const submitPolicy = async () => {
    const reqBody = {
      policyName,
      policyDesc,
      permissions,
      userId: userInfo.user_id,
      status: active ? "Active" : "Inactive",
    };
    if (policyName === "") {
      messageContext.showErrorMessage("Policy Name shouldn't be empty");
      return false;
    }
    let atleastOneSelected = false;
    if (active) {
      Object.keys(permissions).forEach((product) => {
        permissions[product].every((category) => {
          if (!atleastOneSelected) {
            atleastOneSelected = Object.keys(category.permsn_nm).find((x) => {
              return category.permsn_nm[x] === true;
            });
          }
          if (atleastOneSelected) return false;
          return true;
        });
      });
      if (!atleastOneSelected) {
        messageContext.showErrorMessage(
          "Please complete all mandatory information and then click Save"
        );
        return false;
      }
    }
    setLoading(true);
    addPolicyService(reqBody)
      .then((res) => {
        messageContext.showSuccessMessage(res.message || "Successfully Done");
        history.push("/policy-management");
        setLoading(false);
      })
      .catch((err) => {
        messageContext.showErrorMessage(err.message || "Something went wrong");
        setLoading(false);
      });
  };
  const handleChange = (e) => {
    const val = e.target.value;
    if (e.target.id === "policyName") {
      inputAlphaNumeric(e, (v) => {
        setPolicyName(v);
      });
    } else if (e.target.id === "policyDesc") {
      setPolicyDesc(val);
    }
  };
  const filterPermission = (arr) => {
    const helper = {};
    return arr.reduce((r, o) => {
      const key = `${o.ctgy_nm}-${o.feat_nm}`;
      if (!helper[key]) {
        helper[key] = { ...o, permsn_nm: { [o.permsn_nm]: false } };
        r.push(helper[key]);
      } else {
        helper[key].permsn_nm = {
          ...helper[key].permsn_nm,
          [o.permsn_nm]: false,
        };
      }
      return r;
    }, []);
  };
  const fetchPermissions = async () => {
    const permissionsData = await getPolicyPermissions();
    const filteredData = filterPermission(permissionsData);
    const permissionArr = {};
    products.forEach((product) => {
      const filtered = filteredData.filter(
        (x) => x.prod_nm === product.prod_nm
      );
      permissionArr[product.prod_nm] = filtered;
    });
    setPermissions(permissionArr);
  };
  const getProducts = async () => {
    const productsData = await fetchProducts();
    setProducts(productsData);
  };
  const updateData = (childData) => {
    const newArr = { ...permissions, [childData.product]: childData.data };
    console.log("updateData", newArr);
    setPermissions(newArr);
  };
  useEffect(() => {
    fetchPermissions();
  }, [products]);
  useEffect(() => {
    getProducts();
  }, []);
  return (
    <div className="create-policy-wrapper">
      <Box className="top-content">
        <BreadcrumbsUI className="breadcrump" items={breadcrumpItems} />
        <div className="flex top-actions">
          <Switch
            label="Active"
            className="inline-checkbox"
            checked={active}
            onChange={handleActive}
            size="small"
          />
          <ButtonGroup
            alignItems="right"
            buttonProps={[
              {
                label: "Cancel",
                size: "small",
                onClick: () => history.push("/policy-management"),
              },
              {
                label: "Save",
                size: "small",
                disabled: loading,
                onClick: submitPolicy,
              },
            ]}
          />
        </div>
      </Box>
      <Grid container spacing={2}>
        <Grid item xs={3}>
          <Box>
            <div className="flex create-sidebar flexWrap">
              <Typography variant="title1" className="b-font title">
                New Policy
              </Typography>
              <br />
              <TextField
                id="policyName"
                size="small"
                label="Policy Name"
                placeholder="Name your policy"
                onChange={handleChange}
              />
              <TextField
                id="policyDesc"
                size="small"
                label="Policy Description"
                placeholder="Describe your policy"
                rows="18"
                multiline={true}
                minHeight={150}
                sizeAdjustable
                onChange={handleChange}
              />
            </div>
          </Box>
        </Grid>
        <Grid item xs={9} className="products-wrapper">
          <br />
          <Tabs value={currentTab} onChange={handleChangeTab} truncate>
            {products &&
              products.map((product, i) => {
                return <Tab label={product.prod_nm} />;
              })}
          </Tabs>
          <div className="product-content">
            {products &&
              products.map((product, i) => {
                if (currentTab !== i) {
                  return false;
                }
                return (
                  <PermissionTable
                    messageContext={messageContext}
                    title={product.prod_nm}
                    updateData={updateData}
                    data={permissions[product.prod_nm] || []}
                  />
                );
              })}
          </div>
        </Grid>
      </Grid>
    </div>
  );
};

export default CreatePolicy;
