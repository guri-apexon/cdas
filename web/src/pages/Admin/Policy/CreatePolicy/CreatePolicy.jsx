/* eslint-disable no-script-url */
/* eslint-disable react/button-has-type */
import React, { useContext, useEffect, useState, useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import Box from "apollo-react/components/Box";
import { useHistory } from "react-router";
import BreadcrumbsUI from "apollo-react/components/Breadcrumbs";
import Switch from "apollo-react/components/Switch";
import ButtonGroup from "apollo-react/components/ButtonGroup";
import TextField from "apollo-react/components/TextField";
import "./CreatePolicy.scss";
import Typography from "apollo-react/components/Typography";
import Grid from "apollo-react/components/Grid";
import Tab from "apollo-react/components/Tab";
import Tabs from "apollo-react/components/Tabs";
import Modal from "apollo-react/components/Modal";
import Badge from "apollo-react/components/Badge";
import moment from "moment";
import {
  addPolicyService,
  fetchProducts,
  getPolicyPermissions,
} from "../../../../services/ApiServices";
import { MessageContext } from "../../../../components/Providers/MessageProvider";
import PermissionTable from "./PermissionTable";
import { getUserInfo, inputAlphaNumeric } from "../../../../utils";
import {
  formComponentActive,
  hideAlert,
  showAppSwitcher,
  formComponentInActive,
  hideAppSwitcher,
} from "../../../../store/actions/AlertActions";
import AlertBox from "../../../AlertBox/AlertBox";

const ConfirmModal = React.memo(({ open, cancel, closeModal, loading }) => {
  return (
    <Modal
      open={open}
      disableBackdropClick={true}
      onClose={closeModal}
      className="save-confirm"
      variant="warning"
      title="Lose your work?"
      message="All unsaved changes will be lost."
      buttonProps={[
        { label: "Keep editing", onClick: closeModal, disabled: loading },
        { label: "Leave without saving", onClick: cancel, disabled: loading },
      ]}
      id="neutral"
    />
  );
});
const CreatePolicy = () => {
  const dispatch = useDispatch();
  const [active, setActive] = useState(false);
  const [loading, setLoading] = useState(false);
  const [confirm, setConfirm] = useState(false);
  const [currentTab, setCurrentTab] = useState(0);
  const [policyName, setPolicyName] = useState("");
  const [policyDesc, setPolicyDesc] = useState("");
  const [permissions, setPermissions] = useState({});
  const [products, setProducts] = useState([]);
  const messageContext = useContext(MessageContext);
  const userInfo = getUserInfo();
  const history = useHistory();
  const routerHandle = useRef();
  const [targetRoute, setTargetRoute] = useState("");
  const alertStore = useSelector((state) => state.Alert);
  const [isShowAlertBox, setShowAlertBox] = useState(false);

  const unblockRouter = () => {
    dispatch(formComponentInActive());
    dispatch(hideAlert());
    dispatch(hideAppSwitcher());
    if (routerHandle) {
      routerHandle.current();
    }
  };
  const handleActive = (e, checked) => {
    setActive(checked);
  };
  const handleChangeTab = (event, v) => {
    setCurrentTab(v);
  };
  const breadcrumpItems = [
    { href: "javascript:void(0)", onClick: () => history.push("/launchpad") },
    {
      href: "javascript:void(0)",
      title: "Policy Management",
      onClick: () => history.push("/policy-management"),
    },
    {
      title: "Create New Policy",
    },
  ];
  // eslint-disable-next-line consistent-return
  const submitPolicy = async () => {
    const reqBody = {
      policyName,
      policyDesc,
      permissions,
      userId: userInfo.user_id,
      status: active ? "Active" : "Inactive",
      created_on: new Date().toISOString(),
      updated_on: new Date().toISOString(),
    };
    if (policyName === "") {
      messageContext.showErrorMessage("Policy name shouldn't be empty");
      return false;
    }
    if (policyName.length > 255) {
      messageContext.showErrorMessage(
        "Policy name should not allowed to save with more than 255 characters. Max length is 255"
      );
      return false;
    }
    let atleastOneSelected = false;
    if (active) {
      Object.keys(permissions).forEach((product) => {
        permissions[product].every((category) => {
          if (!atleastOneSelected) {
            atleastOneSelected = Object.keys(category.permsn_nm).find((x) => {
              return category.permsn_nm[x] === true;
            });
          }
          if (atleastOneSelected) return false;
          return true;
        });
      });
      if (!atleastOneSelected) {
        messageContext.showErrorMessage(
          "Please complete all mandatory information and then click Save"
        );
        return false;
      }
    }
    setLoading(true);
    addPolicyService(reqBody)
      .then((res) => {
        messageContext.showSuccessMessage(res.message || "Successfully done");
        unblockRouter();
        history.push("/policy-management");
        setLoading(false);
      })
      .catch((err) => {
        messageContext.showErrorMessage(err.message || "Something went wrong");
        setLoading(false);
      });
  };
  const handleChange = (e) => {
    const val = e.target.value;
    if (e.target.id === "policyName") {
      inputAlphaNumeric(e, (v) => {
        setPolicyName(v);
      });
    } else if (e.target.id === "policyDesc") {
      setPolicyDesc(val);
    }
  };
  const cancelCreate = () => {
    unblockRouter();
    if (targetRoute === "") {
      history.push("/policy-management");
    } else {
      history.push(targetRoute);
    }
  };
  const filterPermission = (arr) => {
    if (!arr) return [];
    const helper = {};
    return arr.reduce((r, o) => {
      const key = `${o.ctgy_nm}-${o.feat_nm}`;
      if (!helper[key]) {
        helper[key] = { ...o, permsn_nm: { [o.permsn_nm]: false } };
        r.push(helper[key]);
      } else {
        helper[key].permsn_nm = {
          ...helper[key].permsn_nm,
          [o.permsn_nm]: false,
        };
      }
      return r;
    }, []);
  };
  const fetchPermissions = async () => {
    const permissionsData = await getPolicyPermissions();
    const filteredData = filterPermission(permissionsData);
    const permissionArr = {};
    products.forEach((product) => {
      const filtered = filteredData.filter(
        (x) => x.prod_nm === product.prod_nm
      );
      permissionArr[product.prod_nm] = filtered;
    });
    setPermissions(permissionArr);
  };
  const getProducts = async () => {
    const productsData = await fetchProducts();
    if (productsData) setProducts(productsData);
  };
  const updateData = (childData) => {
    const newArr = { ...permissions, [childData.product]: childData.data };
    setPermissions(newArr);
  };
  const getBadgeCount = (product) => {
    if (!permissions[product]) {
      return 0;
    }
    return permissions[product].filter((category) => {
      return Object.keys(category.permsn_nm).find((x) => {
        return category.permsn_nm[x] === true;
      });
    }).length;
  };
  const closeModal = () => setConfirm(false);
  useEffect(() => {
    fetchPermissions();
  }, [products]);
  useEffect(() => {
    getProducts();
  }, []);

  const keepEditingBtn = () => {
    dispatch(hideAlert());
    setShowAlertBox(false);
  };

  const leavePageBtn = () => {
    dispatch(hideAlert());
    dispatch(showAppSwitcher());
    setShowAlertBox(false);
  };

  useEffect(() => {
    dispatch(formComponentActive());
  }, []);

  useEffect(() => {
    if (alertStore?.showAlertBox) {
      setShowAlertBox(true);
    }
  }, [alertStore]);

  useEffect(() => {
    routerHandle.current = history.block((tr) => {
      setTargetRoute(tr?.pathname);
      setConfirm(true);
      return false;
    });

    return function () {
      /* eslint-disable */
      routerHandle.current()
    };
  });
  return (
    <div className="create-policy-wrapper">
      {isShowAlertBox && (
        <AlertBox cancel={keepEditingBtn} submit={leavePageBtn} />
      )}
      <ConfirmModal
        open={confirm}
        cancel={cancelCreate}
        loading={loading}
        closeModal={closeModal}
      />
      <Box className="top-content">
        <BreadcrumbsUI className="breadcrump" items={breadcrumpItems} />
        <div className="flex top-actions">
          <Switch
            label="Active"
            className="inline-checkbox"
            checked={active}
            onChange={handleActive}
            size="small"
          />
          <ButtonGroup
            alignItems="right"
            buttonProps={[
              {
                label: "Cancel",
                size: "small",
                onClick: setConfirm,
              },
              {
                label: "Save",
                size: "small",
                disabled: loading,
                onClick: submitPolicy,
              },
            ]}
          />
        </div>
      </Box>
      <Grid container spacing={2}>
        <Grid item xs={3}>
          <Box>
            <div className="flex create-sidebar flexWrap">
              <Typography variant="title1" className="b-font title">
                New Policy
              </Typography>
              <br />
              <TextField
                id="policyName"
                size="small"
                label="Policy Name"
                placeholder="Name your policy"
                inputProps={{
                  maxLength: 255,
                }}
                onChange={handleChange}
              />
              <TextField
                id="policyDesc"
                size="small"
                label="Policy Description"
                placeholder="Describe your policy"
                rows="18"
                multiline={true}
                minHeight={150}
                sizeAdjustable
                onChange={handleChange}
              />
            </div>
          </Box>
        </Grid>
        <Grid item xs={9} className="products-wrapper">
          <br />
          <Tabs
            value={currentTab}
            onChange={handleChangeTab}
            truncate
            className="product-tabs"
          >
            {products &&
              products.map((product, i) => {
                return (
                  <Tab
                    key={product.prod_id}
                    disabled={product.active_product === 0}
                    label={
                      // eslint-disable-next-line react/jsx-wrap-multilines
                      <Badge
                        badgeContent={getBadgeCount(product.prod_nm)}
                        max={999}
                      >
                        {product.prod_nm === "Admin"
                          ? "CDAS Admin"
                          : product.prod_nm}
                      </Badge>
                    }
                  />
                );
              })}
          </Tabs>
          <div className="product-content">
            {products &&
              products.map((product, i) => {
                if (currentTab !== i || product.active_product === 0) {
                  return false;
                }
                return (
                  <PermissionTable
                    key={product.prod_id}
                    messageContext={messageContext}
                    title={product.prod_nm}
                    updateData={updateData}
                    data={permissions[product.prod_nm] || []}
                  />
                );
              })}
          </div>
        </Grid>
      </Grid>
    </div>
  );
};

export default CreatePolicy;
