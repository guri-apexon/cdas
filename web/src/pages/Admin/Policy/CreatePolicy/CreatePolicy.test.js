import React from "react";
import { mount } from "enzyme";
import CreatePolicy from "./CreatePolicy";

const wrapper = mount(<CreatePolicy />);
test("AuditLogs component renders", () => {
  setTimeout(() => {
    expect(wrapper).toMatchSnapshot();
  });
});

test("Rendering component", () => {
  expect(wrapper.exists(".create-policy-wrapper")).toBe(true);
});
