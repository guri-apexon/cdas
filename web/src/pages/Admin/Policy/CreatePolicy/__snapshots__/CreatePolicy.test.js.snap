// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuditLogs component renders 1`] = `ReactWrapper {}`;
